// _wepy-cli@1.7.3@wepy-cli/lib/cli/devtool-prompt.js
'use strict';

exports.__esModule = true;
exports.test = exports.upload = exports.preview = exports.login = exports.open = undefined;

var _util = require('../util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var open = exports.open = {
    path: {
        type: 'string',
        required: false,
        message: 'Project path',
        default: ''
    }
};

var login = exports.login = {
    format: {
        type: 'list',
        message: 'Pick a format way',
        choices: [{
            name: 'terminal (命令行输出)',
            value: 'terminal',
            short: 'terminal'
        }, {
            name: 'base64',
            value: 'base64',
            short: 'base64'
        }, {
            name: 'image',
            value: 'image',
            short: 'image'
        }]
    },
    target: {
        type: 'string',
        required: false,
        message: 'Login qrcode output path',
        default: ''
    }
};

var preview = exports.preview = {
    path: {
        type: 'string',
        required: false,
        message: 'Project path',
        default: ''
    },
    format: {
        type: 'list',
        message: 'Pick a format way',
        choices: [{
            name: 'terminal (命令行输出)',
            value: 'terminal',
            short: 'Terminal'
        }, {
            name: 'base64',
            value: 'base64',
            short: 'Base64'
        }, {
            name: 'image',
            value: 'image',
            short: 'Image'
        }]
    },
    target: {
        type: 'string',
        required: false,
        message: 'Preview qrcode output path',
        default: ''
    }
};

var upload = exports.upload = {
    version: {
        type: 'string',
        required: false,
        message: 'Upload Version',
        default: _util2.default.getProjectVersion() || '1.0.0'
    },
    path: {
        type: 'string',
        required: true,
        message: 'Project Path',
        default: _util2.default.currentDir
    },
    desc: {
        type: 'string',
        required: false,
        message: 'Upload Description ',
        default: 'release'
    }
};

var test = exports.test = {
    path: {
        type: 'string',
        required: true,
        message: 'Project Path',
        default: _util2.default.currentDir
    }
};
